Improve primitive object wrapping and refactor PyObject slightly

ilevn opened the pull request:
This PR adds a `Wrappable` interface, which aids with general wrapping of objects of type `T: BaseKycType` into `U: PyObject`.
It also turns no-param getter functions of `PyObject` into member properties for idiomatic access. 

PR is a draft due to pending merge of #3.

Fuyukai commented on src/main/kotlin/interpreter/functions/PyBuiltinFunction.kt:
```diff
@@ -54,10 +54,11 @@ abstract class PyBuiltinFunction(val name: String) : PyFunction(PyBuiltinFunctio
         }
     }
 
-    override fun pyStr(): PyString =
```
No. Everything that's exposed as a python function, is also a Kotlin function. These are the impls of `__str__` and `__repr__`, and as such are exposed as functions.

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyObject.kt:
```diff
@@ -86,12 +59,9 @@ abstract class PyObject() {
-            val d = linkedMapOf<String, PyObject>()
-            return d
-        }
+        val defaultDict: LinkedHashMap<String, PyObject> = linkedMapOf()
```
This is specifically not cached because it should be re-created each time. The default "empty" impl is because of a WIP.

Fuyukai commented on src/main/kotlin/kyc/KycTypes.kt:
```diff
@@ -116,40 +127,61 @@ class KycBoolean private constructor(override val wrapped: Boolean) : BaseKycTyp
 /** An encoded tuple. */
-class KycTuple(override val wrapped: List<BaseKycType>) : BaseKycType()
+class KycTuple(override val wrapped: List<BaseKycType>) : BaseKycType() {
+    override fun wrap(): PyTuple = PyTuple(wrapped.map(PyObject.Companion::wrapKyc))
```
Surely this would just be `wrapped.map { it.wrap() }`?

Fuyukai reviewed (changes requested):
There's probably no need for the companion object wrapKyc function anymore if the interface exists. Just cast to to the interface and wrap directly whenever needed.

ilevn committed:
Address review
- Convert pyStr/pyRepr back to functions to match python signature.
- Remove wrapper companion function.
- Add getter for initialDict to bypass cache.
